src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AddMissingInterpolationToken.cs

```diff
@@ -0,0 +1,107 @@
+// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Analyzer.Utilities;
+using Analyzer.Utilities.Extensions;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Microsoft.CodeQuality.Analyzers.QualityGuidelines
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+    public sealed class AddMissingInterpolationTokenAnalyzer : DiagnosticAnalyzer
+    {
+        internal const string RuleId = "CA2017";
+
+        internal static DiagnosticDescriptor Rule = DiagnosticDescriptorHelper.Create(RuleId,
+            new LocalizableResourceString(nameof(MicrosoftCodeQualityAnalyzersResources.AddMissingInterpolationTokenTitle), MicrosoftCodeQualityAnalyzersResources.ResourceManager, typeof(MicrosoftCodeQualityAnalyzersResources)),
+            new LocalizableResourceString(nameof(MicrosoftCodeQualityAnalyzersResources.AddMissingInterpolationTokenMessage), MicrosoftCodeQualityAnalyzersResources.ResourceManager, typeof(MicrosoftCodeQualityAnalyzersResources)),
+            DiagnosticCategory.Reliability,
+            RuleLevel.IdeSuggestion,
+            description: null,
+            isPortedFxCopRule: false,
+            isDataflowRule: false);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext analysisContext)
+        {
+            analysisContext.EnableConcurrentExecution();
+            analysisContext.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            analysisContext.RegisterOperationAction(context =>
+            {
+                var literalOperation = (ILiteralOperation)context.Operation;
+
+                if (!literalOperation.ConstantValue.HasValue ||
+                    literalOperation.ConstantValue.Value is not string stringText)
+                {
+                    return;
+                }
+
+                var parts = GetInterpolationParts(stringText);
+                if (parts.All(s => IsValidInterpolationPart(s, literalOperation.SemanticModel, literalOperation.Syntax.SpanStart)))
+                {
+                    context.ReportDiagnostic(literalOperation.CreateDiagnostic(Rule));
+                }
+            }, OperationKind.Literal);
+        }
+
+        private static bool IsValidInterpolationPart(string s, SemanticModel semanticModel, int position)
```

Comment:
Looking at https://docs.microsoft.com/dotnet/csharp/language-reference/tokens/interpolated#structure-of-an-interpolated-string I think that your validation algorithm is wrong.

Reply:
@Evangelink The implementation currently handles the very simple cases only. I think I'd need some better way to get this information using some compiler APIs. Do you have ideas?